File: zig/bench.zig
const std = @import("std");

const N_NODES: usize = 32;
const BLOOM_WORDS: usize = 94;
const SKETCH_D: usize = 7;
const SKETCH_W: usize = 1000;
const N_RUNS: usize = 2_000_000;

fn gcounter_merge_simd(dst: *[N_NODES]u64, src: *const [N_NODES]u64) void {
    const Vec16 = @Vector(16, u64);
    const half = N_NODES / 2;
    const dst_lo: Vec16 = dst[0..half].*;
    const src_lo: Vec16 = src[0..half].*;
    dst[0..half].* = @max(dst_lo, src_lo);
    const dst_hi: Vec16 = dst[half..].*;
    const src_hi: Vec16 = src[half..].*;
    dst[half..].* = @max(dst_hi, src_hi);
}

fn gcounter_merge_scalar(dst: *[N_NODES]u64, src: *const [N_NODES]u64) void {
    for (dst, src) |*d, s| d.* = @max(d.*, s);
}

fn bloom_merge_simd(dst: *[BLOOM_WORDS]u64, src: *const [BLOOM_WORDS]u64) void {
    const Vec8 = @Vector(8, u64);
    comptime var ci: usize = 0;
    inline while (ci + 8 <= BLOOM_WORDS) : (ci += 8) {
        const d: Vec8 = dst[ci..][0..8].*;
        const s: Vec8 = src[ci..][0..8].*;
        dst[ci..][0..8].* = d | s;
    }
    var i: usize = ci;
    while (i < BLOOM_WORDS) : (i += 1) dst[i] |= src[i];
}

fn bloom_merge_scalar(dst: *[BLOOM_WORDS]u64, src: *const [BLOOM_WORDS]u64) void {
    for (dst, src) |*d, s| d.* |= s;
}

fn sketch_merge_simd(dst: *[SKETCH_D][SKETCH_W]u64, src: *const [SKETCH_D][SKETCH_W]u64) void {
    const Vec8 = @Vector(8, u64);
    for (0..SKETCH_D) |d| {
        comptime var w: usize = 0;
        inline while (w + 8 <= SKETCH_W) : (w += 8) {
            const dv: Vec8 = dst[d][w..][0..8].*;
            const sv: Vec8 = src[d][w..][0..8].*;
            dst[d][w..][0..8].* = @max(dv, sv);
        }
        while (w < SKETCH_W) : (w += 1) dst[d][w] = @max(dst[d][w], src[d][w]);
    }
}

fn sketch_merge_scalar(dst: *[SKETCH_D][SKETCH_W]u64, src: *const [SKETCH_D][SKETCH_W]u64) void {
    for (0..SKETCH_D) |d| {
        for (0..SKETCH_W) |w| dst[d][w] = @max(dst[d][w], src[d][w]);
    }
}

var gc_a: [N_NODES]u64 = undefined;
var gc_b: [N_NODES]u64 = undefined;
var bl_a: [BLOOM_WORDS]u64 = undefined;
var bl_b: [BLOOM_WORDS]u64 = undefined;
var sk_a: [SKETCH_D][SKETCH_W]u64 = undefined;
var sk_b: [SKETCH_D][SKETCH_W]u64 = undefined;

fn init() void {
    for (0..N_NODES) |i| { gc_a[i] = @intCast(i * i); gc_b[i] = 100; }
    for (0..BLOOM_WORDS) |i| { bl_a[i] = @intCast(i * i * i); bl_b[i] = 0xAAAAAAAAAAAAAAAA; }
    for (0..SKETCH_D) |d| for (0..SKETCH_W) |w| {
        sk_a[d][w] = @intCast((d + 1) * (w + 1));
        sk_b[d][w] = @intCast((d + 2) * (w + 3));
    };
}

pub fn main() !void {
    init();
    const w = std.debug.print;

    w("╔═══════════════════════════════════════════════════════════════╗\n", .{});
    w("║  CRDT MERGE — Zig 0.16 (Zen 5 + AVX-512)                   ║\n", .{});
    w("╚═══════════════════════════════════════════════════════════════╝\n\n", .{});

    // G-Counter
    run("G-Counter: @Vector(16,u64) SIMD max", struct { fn f() void { gcounter_merge_simd(&gc_a, &gc_b); } }.f);
    run("G-Counter: scalar loop @max", struct { fn f() void { gcounter_merge_scalar(&gc_a, &gc_b); } }.f);
    w("\n", .{});

    // Bloom
    run("Bloom: @Vector(8,u64) SIMD OR", struct { fn f() void { bloom_merge_simd(&bl_a, &bl_b); } }.f);
    run("Bloom: scalar loop |=", struct { fn f() void { bloom_merge_scalar(&bl_a, &bl_b); } }.f);
    w("\n", .{});

    // Sketch
    run("Sketch: @Vector(8) SIMD max (7K)", struct { fn f() void { sketch_merge_simd(&sk_a, &sk_b); } }.f);
    run("Sketch: scalar loop max (7K)", struct { fn f() void { sketch_merge_scalar(&sk_a, &sk_b); } }.f);

    w("\n--- Zig @Vector SIMD analysis ---\n", .{});
    w("  @Vector(16, u64): maps to vpmaxuq (AVX-512) on Zen 5\n", .{});
    w("  @Vector(8, u64):  maps to vor (AVX-512) for Bloom OR\n", .{});
    w("  comptime inline: loop unrolling at compile time, zero overhead\n", .{});
}

const linux = std.os.linux; var _ts: linux.timespec = undefined;
fn clock_ns() i128 {
    _ = linux.clock_gettime(.MONOTONIC, &_ts); // CLOCK_MONOTONIC
    return @as(i128, _ts.sec) * 1_000_000_000 + _ts.nsec;
}

fn run(comptime label: []const u8, comptime f: fn () void) void {
    for (0..1000) |_| f();
    const start = clock_ns();
    for (0..N_RUNS) |_| f();
    const elapsed = clock_ns() - start;
    const ns: f64 = @floatFromInt(@divTrunc(elapsed, @as(i128, @intCast(N_RUNS))));
    if (ns > 0) {
        std.debug.print("  {s:<45} {:>8.1} ns  {:>12.0} ops/s\n", .{ label, ns, 1e9 / ns });
    } else {
        std.debug.print("  {s:<45}    <1.0 ns  (optimized away?)\n", .{label});
    }
}

File: go/bench.go
package main

import (
	"fmt"
	"time"
	"unsafe"
)

const (
	N_NODES      = 32
	BLOOM_WORDS  = 94
	SKETCH_D     = 7
	SKETCH_W     = 1000
	N_RUNS       = 2_000_000
)

// ---- G-Counter merge: element-wise max ----
func gcounterMergeScalar(dst, src []uint64) {
	for i := range dst {
		if src[i] > dst[i] {
			dst[i] = src[i]
		}
	}
}

func gcounterMergeUnsafe(dst, src []uint64) {
	d := unsafe.Slice((*uint64)(unsafe.Pointer(&dst[0])), len(dst))
	s := unsafe.Slice((*uint64)(unsafe.Pointer(&src[0])), len(src))
	for i := range d {
		if s[i] > d[i] {
			d[i] = s[i]
		}
	}
}

// ---- Bloom filter merge: bitwise OR ----
func bloomMergeScalar(dst, src []uint64) {
	for i := range dst {
		dst[i] |= src[i]
	}
}

func bloomMergeUnrolled4(dst, src []uint64) {
	n := len(dst)
	i := 0
	for i+3 < n {
		dst[i] |= src[i]
		dst[i+1] |= src[i+1]
		dst[i+2] |= src[i+2]
		dst[i+3] |= src[i+3]
		i += 4
	}
	for i < n {
		dst[i] |= src[i]
		i++
	}
}

func bloomMergeUnsafe(dst, src []uint64) {
	d := unsafe.Slice((*uint64)(unsafe.Pointer(&dst[0])), len(dst))
	s := unsafe.Slice((*uint64)(unsafe.Pointer(&src[0])), len(src))
	for i := range d {
		d[i] |= s[i]
	}
}

// ---- Count-Min Sketch merge: element-wise max ----
func sketchMergeScalar(dst, src [][SKETCH_W]uint64) {
	for d := range dst {
		for w := 0; w < SKETCH_W; w++ {
			if src[d][w] > dst[d][w] {
				dst[d][w] = src[d][w]
			}
		}
	}
}

func sketchMergeFlat(dst, src []uint64) {
	for i := range dst {
		if src[i] > dst[i] {
			dst[i] = src[i]
		}
	}
}

// ---- Eisenstein norm ----
func e12Norm(a, b int32) int64 {
	la, lb := int64(a), int64(b)
	return la*la - la*lb + lb*lb
}

func bench(name string, f func()) {
	// Warmup
	for i := 0; i < 1000; i++ {
		f()
	}
	start := time.Now()
	for i := 0; i < N_RUNS; i++ {
		f()
	}
	elapsed := time.Since(start)
	nsPer := float64(elapsed.Nanoseconds()) / float64(N_RUNS)
	opsPerS := 1e9 / nsPer
	fmt.Printf("  %-45s %8.1f ns  %12.0f ops/s\n", name, nsPer, opsPerS)
}

func main() {
	fmt.Println("╔═══════════════════════════════════════════════════════════════╗")
	fmt.Println("║  CRDT MERGE — Go 1.24 (Zen 5)                               ║")
	fmt.Println("╚═══════════════════════════════════════════════════════════════╝")
	fmt.Printf("  Runs: %d, Nodes: %d, Bloom: %d words\n\n", N_RUNS, N_NODES, BLOOM_WORDS)

	// G-Counter
	gcA := make([]uint64, N_NODES)
	gcB := make([]uint64, N_NODES)
	for i := range gcA {
		gcA[i] = uint64(i * i)
		gcB[i] = 100
	}

	bench("G-Counter: range loop max", func() { gcounterMergeScalar(gcA, gcB) })
	bench("G-Counter: unsafe ptr max", func() { gcounterMergeUnsafe(gcA, gcB) })

	fmt.Println()

	// Bloom
	blA := make([]uint64, BLOOM_WORDS)
	blB := make([]uint64, BLOOM_WORDS)
	for i := range blA {
		blA[i] = uint64(i * i * i)
		blB[i] = 0xAAAAAAAAAAAAAAAA
	}

	bench("Bloom: range loop |=", func() { bloomMergeScalar(blA, blB) })
	bench("Bloom: unrolled ×4 |=", func() { bloomMergeUnrolled4(blA, blB) })
	bench("Bloom: unsafe ptr |=", func() { bloomMergeUnsafe(blA, blB) })

	fmt.Println()

	// Sketch
	var skA [SKETCH_D][SKETCH_W]uint64
	var skB [SKETCH_D][SKETCH_W]uint64
	for d := 0; d < SKETCH_D; d++ {
		for w := 0; w < SKETCH_W; w++ {
			skA[d][w] = uint64((d + 1) * (w + 1))
			skB[d][w] = uint64((d + 2) * (w + 3))
		}
	}

	bench("Sketch: 2D loop max (7K cells)", func() { sketchMergeScalar(skA[:], skB[:]) })

	// Flat slice version
	flatA := make([]uint64, SKETCH_D*SKETCH_W)
	flatB := make([]uint64, SKETCH_D*SKETCH_W)
	for i := range flatA {
		flatA[i] = uint64(i + 1)
		flatB[i] = uint64(i + 3)
	}
	bench("Sketch: flat slice max (7K cells)", func() { sketchMergeFlat(flatA, flatB) })

	fmt.Println()
	fmt.Println("--- Go analysis ---")
	fmt.Println("  Go has no SIMD intrinsics (no @Vector like Zig)")
	fmt.Println("  unsafe.Slice bypasses bounds checking but doesn't add vectorization")
	fmt.Println("  GC pauses not visible in micro-benchmarks (<1µs per merge)")
	fmt.Println("  Go's strength: goroutine-based parallel merge, not single-thread speed")

	// Parallel merge test
	fmt.Println()
	fmt.Println("--- Parallel merge (4 goroutines, 10K states) ---")
	type State struct {
		counts []uint64
		bloom  []uint64
	}
	states := make([]State, 10000)
	for i := range states {
		states[i].counts = make([]uint64, N_NODES)
		states[i].bloom = make([]uint64, BLOOM_WORDS)
		for j := range states[i].counts {
			states[i].counts[j] = uint64(i + j)
		}
		for j := range states[i].bloom {
			states[i].bloom[j] = 0xAA
		}
	}
	src := State{
		counts: make([]uint64, N_NODES),
		bloom:  make([]uint64, BLOOM_WORDS),
	}
	for i := range src.counts { src.counts[i] = 999 }
	for i := range src.bloom { src.bloom[i] = 0xFF }

	// Sequential
	start := time.Now()
	for i := range states {
		gcounterMergeScalar(states[i].counts, src.counts)
		bloomMergeScalar(states[i].bloom, src.bloom)
	}
	seqMs := time.Since(start).Milliseconds()

	// Parallel (4 goroutines)
	done := make(chan bool, 4)
	chunk := len(states) / 4
	start = time.Now()
	for g := 0; g < 4; g++ {
		go func(from, to int) {
			for i := from; i < to; i++ {
				gcounterMergeScalar(states[i].counts, src.counts)
				bloomMergeScalar(states[i].bloom, src.bloom)
			}
			done <- true
		}(g*chunk, (g+1)*chunk)
	}
	for g := 0; g < 4; g++ { <-done }
	parMs := time.Since(start).Milliseconds()

	fmt.Printf("  Sequential: %d ms, Parallel (4 goroutines): %d ms, Speedup: %.1fx\n",
		seqMs, parMs, float64(seqMs)/float64(max(parMs, 1)))
}
